compiler/rustc_data_structures/src/sharded.rs

```diff
@@ -75,9 +59,24 @@ impl<T> Sharded<T> {
     }
 
     #[inline]
-    pub fn get_shard_by_index(&self, i: usize) -> &Lock<T> {
-        // SAFETY: The index get ANDed with the mask, ensuring it is always inbounds.
-        unsafe { &self.shards.get_unchecked(i & self.mask()).0 }
+    pub fn get_shard_by_index(&self, _i: usize) -> &Lock<T> {
+        match self {
+            Self::Single(single) => &single,
+            #[cfg(parallel_compiler)]
+            Self::Shards(shards) => {
+                // SAFETY: The index gets ANDed with the shard mask, ensuring it is always inbounds.
+                unsafe { &shards.get_unchecked(_i & (SHARDS - 1)).0 }
+            }
+        }
+    }
+
+    #[inline]
+    fn count(&self) -> usize {
+        match self {
+            Self::Single(..) => 1,
+            #[cfg(parallel_compiler)]
+            Self::Shards(..) => SHARDS,
+        }
     }
 
     pub fn lock_shards(&self) -> Vec<LockGuard<'_, T>> {
```

Comment:
I think we discriminant the enum here and `try_lock_shards` below so that we don't need `fn count` any more